piptools/repositories/pypi.py

```diff
@@ -236,15 +236,23 @@ def _get_project(self, ireq):
 
         See https://warehouse.readthedocs.io/api-reference/json/
         """
-        url = "{url}/{name}/json".format(url=PyPI.pypi_url, name=ireq.name)
-        try:
-            data = self.session.get(url).json()
-        except (ValueError, HTTPError) as e:
-            log.error(
-                "Fetch package info from PyPI failed: {url}: {e}".format(url=url, e=e)
-            )
-            return None
-        return data
+        package_indexes = [
+            PackageIndex(url=index_url, file_storage_domain="")
+            for index_url in self.finder.search_scope.index_urls
+        ]
+        for package_index in package_indexes:
+            url = "{url}/{name}/json".format(url=package_index.pypi_url, name=ireq.name)
+            try:
+                data = self.session.get(url).json()
```

Comment:
we should probably remember which URLs support the PyPI JSON api

Reply:
How would we know that JSON API of an extra index URL is supported?

Reply:
if this URL doesn't 404

Reply:
If it raises 404 then it falls back to the files hash checking.

Reply:
Perhaps 404 should be handled explicitly

Reply:
but there's no point re-requesting the API from indexes that we've already tried

Reply:
Good call. But what if https://custom.pypi-server.com/pypi/my-package/json raise 404 because there is no `my-package` on the server, but it supports JSON API?

Reply:
when you fall back to calculating hashes directly you'll know it was because the index doesn't support the api

Reply:
Sorry, I might be missing your point, but I'll try to describe mine in detail. 

Using this heuristic, we can't be sure that JSON API is not supported. Consider the following case:

- package1 from https://pypi1-server/pypi/package1/json
- package2 from https://pypi2-server/pypi/package2/json
- package3 from https://pypi1-server/pypi/package1/json

Suppose we run the following command:
```console
$ pip-compile --index-url=https://pypi1-server/ --extra-index-url=https://pypi2-server/
```

Step 1. When pip-tools processes `package1` everything is okay because it hits https://pypi1-server/pypi/package1/json which responses 200. 

Step2. After,  pip-tools compiles `package2` and hit the first index URL, which is https://pypi1-server/pypi/package2/json. But this URL raises 404 because the package is on pypi2-server. So it has to go to the second index URL, which is https://pypi2-server/pypi/package2/json, which responses 200.

Step3. Then pip-tools processes `package3 `the same as in *step 1* with the first index URL.


> How would we know that JSON API of a custom index URL is supported?
> If this URL doesn't 404.

Considering the above statement the test case will be failed on *step 3* because `pypi1-server` will be marked as "no supported JSON API" on *step 2* because pip-tools couldn't find package2 on `pypi1-server` and it responded 404.

Am I missing something?

Reply:
if the simple index and the JSON API 404s then you don't know
if the simple index 200s and the JSON API 404s then you know the API is unsupported

Reply:
I see your point. Yes, that would work. Assuming that suggestion is to hit less the PyPI servers now we have to hit the servers twice on 404. So what's the point?

Reply:
> You have to hit it twice to download the hashes from indexes that don't
support the JSON API

Sorry, I don't understand what are you trying to achieve by that. Could you elaborate?

Reply:
If you think this should be fixed anyway please feel free to open an issue/PR. Thank you for the help!